### PR TITLE
Hopper (`sm90`) support

### PR DESCRIPTION
**Pull request by eqy**
Essentially a followup of #87436

CC @xwang233 @ptrblck

### CODE REVIEW

**Review by xwang233 — APPROVED** [review]
LGTM

**Review by malfet — APPROVED** [review]
LGTM, but don't we want some sort of gating here, i.e. if CUDA version is less than 11.8, it should not enable sm_90?

**Comment by eqy** [comment]
I assume the gating is done here: https://github.com/pytorch/pytorch/blob/0d60397455baf0f9844d6f93b84628e7649cd3da/torch/utils/cpp_extension.py#L1754 and with the assumption in the above comment that the `nvcc` version used for the extension build should match that of the PyTorch build.

File: cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake [hunk 1]
```diff
@@ -105,8 +105,11 @@ endif()
 
 if(NOT CUDA_VERSION VERSION_LESS "11.8")
   list(APPEND CUDA_KNOWN_GPU_ARCHITECTURES "Ada")
+  list(APPEND CUDA_KNOWN_GPU_ARCHITECTURES "Hopper")
   list(APPEND CUDA_COMMON_GPU_ARCHITECTURES "8.9")
+  list(APPEND CUDA_COMMON_GPU_ARCHITECTURES "9.0" "9.0+PTX")
```
Comment (yaox12) [inline cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake:110]: `"9.0+PTX"` should not be appended here. We should update line 116 instead.

Also, `"8.6+PTX"` should not be appended in line 97. We should add 
```
list(APPEND CUDA_COMMON_GPU_ARCHITECTURES "8.6+PTX")
```
in line 103.

File: cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake [hunk 2]
```diff
@@ -105,8 +105,11 @@ endif()
...
   list(APPEND CUDA_ALL_GPU_ARCHITECTURES "8.9")
+  list(APPEND CUDA_ALL_GPU_ARCHITECTURES "9.0")
 
   if(CUDA_VERSION VERSION_LESS "12.0")
     set(CUDA_LIMIT_GPU_ARCHITECTURE "9.0")
```
Comment (yaox12) [inline cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake:115]: I think `CUDA_LIMIT_GPU_ARCHITECTURE` should be larger than the latest supported arch.

File: cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake [hunk 3]
```diff
@@ -105,8 +105,11 @@ endif()
 
 if(NOT CUDA_VERSION VERSION_LESS "11.8")
   list(APPEND CUDA_KNOWN_GPU_ARCHITECTURES "Ada")
+  list(APPEND CUDA_KNOWN_GPU_ARCHITECTURES "Hopper")
```
Comment (yaox12) [inline cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake:108]: We should add bin/ptx for Ada and Hopper in 
https://github.com/pytorch/pytorch/blob/95fc0bcaaddc2d24e8759f24dbefa789d04e9e42/cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake#L248-L250

**Pull request closed without merging**